Index: libtorrent/src/storage.cpp
===================================================================
--- libtorrent/src/storage.cpp	(revision 3507)
+++ libtorrent/src/storage.cpp	(working copy)
@@ -2228,7 +2228,7 @@
 				&& m_storage->error() != error_code(ENOENT, get_posix_category()))
 #endif
 			{
-				m_piece_data.reset();
+				m_piece_data.clear();
 				return -1;
 			}
 			return 1;
Index: libtorrent/src/policy.cpp
===================================================================
--- libtorrent/src/policy.cpp	(revision 3507)
+++ libtorrent/src/policy.cpp	(working copy)
@@ -659,7 +659,7 @@
 #endif
 			i->second.inet_as = ses.lookup_as(as);
 #endif
-			i->source = peer_info::incoming;
+			i->second.source = peer_info::incoming;
 		}
 	
 		c.set_peer_info(&i->second);
