--- a/twisted/internet/_glibbase.py
+++ b/twisted/internet/_glibbase.py
@@ -22,7 +22,8 @@ from twisted.internet.abstract import FileDescriptor
 from twisted.internet.interfaces import IReactorFDSet, IReadDescriptor, IWriteDescriptor
 from twisted.python import log
 from twisted.python.monkey import MonkeyPatcher
-from ._signals import _IWaker, _UnixWaker
+from twisted.python.runtime import platform
+from ._signals import _IWaker, _Waker


 def ensureNotImported(moduleNames, errorMessage, preventImports=[]):
@@ -52,7 +53,7 @@ def ensureNotImported(moduleNames, errorMessage, preventImports=[]):
         sys.modules[name] = None


-class GlibWaker(_UnixWaker):
+class GlibWaker(_Waker):
     """
     Run scheduled events after waking up.
     """
@@ -209,6 +210,11 @@ class GlibReactorBase(posixbase.PosixReactorBase, posixbase._PollLikeMixin):
         else:
             fileno = source
             wrapper = callback
+
+        if platform.isWindows():
+            # GLib can't watch socket fds directly on Windows; needs IOChannel.
+            fileno = self._glib.IOChannel.win32_new_socket(fileno)
+
         return self._glib.io_add_watch(
             fileno,
             self._glib.PRIORITY_DEFAULT_IDLE,
